feat(dot/rpc): implement RPC method state_queryStorageAt

edwardmack opened the pull request:
## Changes
- Implement code for RPC method `state_queryStorageAt` which returns storage values for given keys at best block or given block hash if `at` parameter is used.
- This fixes error `rpc: can't find method "state.QueryStorageAt"` that occurs when using polkadot.js/api methods in `api.query` package.

## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
go test github.com/ChainSafe/gossamer/dot/rpc/modules -run ^TestStateModuleQueryStorageAt$ -v
```

## Issues
closes: #3182 

## Primary Reviewer

<!-- Tag a code owner to review your PR -->

@timwu20

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -76,6 +76,12 @@ type StateStorageQueryRangeRequest struct {
 	EndBlock   common.Hash `json:"block"`
 }
 
+// StateStorageQueryAtRequest holds json fields
+type StateStorageQueryAtRequest struct {
```
I think `StateStorageQueryAtRequest` can be unexported

kishansagathiya replied:
That would apply to its properties as well.

I don't see any variable of type StateStorageQueryAtRequest getting used by scale. Because scale requires structs and properties to be exported.

qdm12 replied:
I think they are exported for the encoding/json encoder to work right?

edwardmack replied:
Yes, they are exported for encoding/json encoder to work.

edwardmack replied:
Yes, we're using Gorilla web toolkit to expose functions through RPC, Gorilla exposes functions with a signature of containing exported structs for request and response.  I tried changing StateStorageQueryAtRequest to un-exported and the RPC function was no longer accessible.

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
 	return nil
 }
 
+// QueryStorageAt queries historical storage entries (by key) at the best block if the given block hash is nil
```
```suggestion
// QueryStorageAt queries historical storage entries (by key) at the block hash given or 
// the best block if the given block hash is nil
```

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
 
+// QueryStorageAt queries historical storage entries (by key) at the best block if the given block hash is nil
+func (sm *StateModule) QueryStorageAt(
+	_ *http.Request, req *StateStorageQueryAtRequest, res *[]StorageChangeSetResponse) error {
```
nit `req` -> `request` and `res` -> `response` (see comment below to avoid name conflict)

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
+// QueryStorageAt queries historical storage entries (by key) at the best block if the given block hash is nil
+func (sm *StateModule) QueryStorageAt(
+	_ *http.Request, req *StateStorageQueryAtRequest, res *[]StorageChangeSetResponse) error {
+	var atBlockHash common.Hash
+	if req.At.IsEmpty() {
+		atBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		atBlockHash = req.At
+	}
```
nit a tiny bit shorter

```suggestion
	atBlockHash := req.At
	if atBlockHash.IsEmpty() {
		atBlockHash = sm.blockAPI.BestBlockHash()
	}
```

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
+		changes = append(changes, [2]*string{stringPtr(key), hexValue})
+	}
+
+	response = append(response, StorageChangeSetResponse{
+		Block:   &atBlockHash,
+		Changes: changes,
+	})
+
+	*res = response
```
nit change to directly be:

```go
	res = new([]StorageChangeSetResponse)
	*res = []StorageChangeSetResponse{{
		Block:   &atBlockHash,
		Changes: changes,
	}}
```

that way you can also remove the init `response := make([]StorageChangeSetResponse, 0, 1)` above and avoid name conflicts between `res` and `response`.

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
+			hexValue = stringPtr(common.BytesToHex(value))
+		}
+
+		changes = append(changes, [2]*string{stringPtr(key), hexValue})
```
nit since there is no `continue` clause, you can set it by index directly and create changes with

```go
changes := make([][2]*string, len(req.Keys))
```

and then have

```go
for i, key := range req.Keys {
    // ...
    changes[i] = [2]*string{stringPtr(key), hexValue})
}
```

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+			fields: fields{
+				storageAPIBuilder: func(ctrl *gomock.Controller) StorageAPI {
+					mockStorageAPI := NewMockStorageAPI(ctrl)
+					mockStorageAPI.EXPECT().GetStorageByBlockHash(&common.Hash{2}, []byte{1, 2, 3}).Return([]byte{1, 1, 1}, nil)
```
nit split since it's a bit of a long line 😉 

```suggestion
					mockStorageAPI.EXPECT().GetStorageByBlockHash(&common.Hash{2}, []byte{1, 2, 3}).
						Return([]byte{1, 1, 1}, nil)
```

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+		blockAPIBuilder   func(ctrl *gomock.Controller) BlockAPI
+	}
+	type args struct {
+		in0 *http.Request
```
Since `in0` is always ignored, just remove it and always pass nil in the subtest code?

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+	errTest := errors.New("test error")
+
+	type fields struct {
+		storageAPIBuilder func(ctrl *gomock.Controller) StorageAPI
+		blockAPIBuilder   func(ctrl *gomock.Controller) BlockAPI
```
nit let's return mock types (better to not return interfaces generally)

```go
		storageAPIBuilder func(ctrl *gomock.Controller) *MockStorageAPI
		blockAPIBuilder   func(ctrl *gomock.Controller) *MockBlockAPI
```

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+	}
+	type args struct {
+		in0 *http.Request
+		req *StateStorageQueryAtRequest
```
nit rename `req` -> `request`

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+	tests := map[string]struct {
+		fields    fields
+		args      args
+		errRegexp string
```
no need for a regexp, the only error case is predictable (`getting value by block hash: test error`)

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+		fields    fields
+		args      args
+		errRegexp string
+		exp       []StorageChangeSetResponse
```
nit `exp` -> `response` or `expectedResponse` would be nicer 😉

qdm12 commented on dot/rpc/modules/state_test.go:
```diff
@@ -1129,3 +1129,147 @@ func TestStateModuleQueryStorage(t *testing.T) {
+				storageAPI: tt.fields.storageAPIBuilder(ctrl),
+				blockAPI:   tt.fields.blockAPIBuilder(ctrl),
+			}
+			res := []StorageChangeSetResponse{}
```
nit `res` -> `response` would be nicer 😉

qdm12 commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
+			return fmt.Errorf("getting value by block hash: %w", err)
+		}
+		var hexValue *string
+		if len(value) > 0 {
```
should we differentiate `[]byte{}` and `[]byte(nil)`?? This comes from the trie and the empty byte slice means it's set. The `nil` byte slice means the value doesn't exist though. I would tend to think we should return `0x` in case the value returned is an empty-non-nil byte slice.

edwardmack replied:
Good point, I've updated this function and `state_QueryStorage` to handle `[]byte{}` and `[]byte(nil)` and added tests for both conditions.

edwardmack commented on dot/rpc/modules/state.go:
```diff
@@ -489,6 +500,42 @@ func (sm *StateModule) QueryStorage(
+		changes[i] = [2]*string{stringPtr(key), hexValue}
+	}
+
+	*response = []StorageChangeSetResponse{{
```
I think this is because `response` is initialized since it is a function argument.

edwardmack commented:
This PR has been moved to PR #3191 which contains only code related to implementing state_queryStorageAt (not other merged code from other branches).  @qdm12 please let me know if the remaining comments on this can be resolved, I believe I've addressed all comments in the new PR.

edwardmack commented on cmd/gossamer/config_test.go:
```diff
@@ -845,7 +845,7 @@ func TestGlobalNodeName_WhenNodeAlreadyHasStoredName(t *testing.T) {
 	cfg.Global.Name = globalName
 
-	runtimeFilePath, err := runtime.GetRuntime(context.Background(), runtime.NODE_RUNTIME)
+	runtimeFilePath, err := runtime.GetRuntime(context.Background(), runtime.WESTEND_RUNTIME_v0929)
```
This PR has been moved to now PR #3191 which fixes the rebase issue.
